File: web/src/services/comments.ts
import { notifyTaskChanged } from '../lib/taskChanges'

import { ApiError } from './auth'

/**
 * Comments client, mirroring the backend JSON shapes of `internal/comments` and
 * the comment routes of `internal/photoapi` and `internal/phototaskapi`. A comment
 * is one short plain-text note by one user on one subject: a photograph — the
 * family conversation around a picture — or a task, where it is the answer to the
 * question the task asks.
 *
 * Which of the two a thread belongs to is a {@link CommentSubject}; everything
 * below it is identical, because server-side it is one table and one set of rules.
 *
 * Two properties of the backend shape the whole client:
 *
 *   - **Bodies are plain text.** Nothing is parsed, rendered or sanitised
 *     server-side, so whatever displays a body must escape it (React does, as long
 *     as nobody reaches for `dangerouslySetInnerHTML`).
 *   - **Writing is open to every signed-in role, viewers included.** The create
 *     route is guarded by `RequireAuth`, not `RequireWrite` — the one documented
 *     exception to the read-only rule — so this client never gates on `canWrite`.
 *
 * Every call throws {@link ApiError} on a non-OK response so callers can branch on
 * `status`: 403 (someone else's comment), 404 (already deleted), 429 (the per-user
 * rate limit) and 400 (empty or over-long body) all mean different things to a reader.
 */

const API_BASE = '/api/v1'

/**
 * The longest comment body the backend accepts, in characters (runes, not bytes),
 * mirroring `comments.MaxBodyLen`. The composer caps its input at the same number
 * so an over-long body is prevented rather than rejected after the fact.
 */
export const MAX_COMMENT_LENGTH = 2000

/** What a comment thread hangs off: a photograph, or a task. */
export type CommentSubjectKind = 'photo' | 'task'

/** The one thing a thread is about — the kind, and the uid of that row. */
export interface CommentSubject {
  kind: CommentSubjectKind
  uid: string
}

/** Names the thread of a photograph. */
export function photoSubject(uid: string): CommentSubject {
  return { kind: 'photo', uid }
}

/** Names the thread of a task. */
export function taskSubject(uid: string): CommentSubject {
  return { kind: 'task', uid }
}

/**
 * One stored comment as read back from the API (`comments.Comment`).
 *
 * Exactly one of `photo_uid` and `task_uid` is set, matching the row's one subject.
 *
 * `author_uid` and `author_name` are **empty strings** for a comment whose author's
 * account has since been deleted: the row survives authorless, and nobody may edit
 * it any more. Renderers must therefore not assume a name is present.
 */
export interface Comment {
  uid: string
  /** The photograph this comment is on; absent on a task thread. */
  photo_uid?: string
  /** The task this comment is on; absent on a photo thread. */
  task_uid?: string
  author_uid: string
  /** The author's display name (falling back to the username), resolved server-side. */
  author_name: string
  /**
   * The cover photo of the person the author's account says it is. It predates
   * profile pictures and the thread no longer draws it: `PersonAvatar` asks
   * `GET /users/{uid}/avatar` with `author_uid`, which resolves the whole chain
   * (an uploaded picture, a picked photo, the linked person's face) rather than
   * this one link of it.
   */
  author_photo_uid?: string
  body: string
  created_at: string
  /** Set once the author has rewritten the body; absent on a never-edited comment. */
  edited_at?: string
}

/** Response body of the thread endpoint of either subject. */
interface CommentListResponse {
  comments: Comment[]
}

/** Standard backend error envelope shared by every API group. */
interface ErrorBody {
  error?: string
}

/** Extracts the backend error message from a non-OK response, if present. */
async function readErrorMessage(res: Response): Promise<string> {
  try {
    const body = (await res.json()) as ErrorBody
    if (typeof body.error === 'string' && body.error !== '') {
      return body.error
    }
  } catch {
    // Body was empty or not JSON; fall back to the status text below.
  }
  return res.statusText || `request failed: ${res.status}`
}

/** Issues a request against the comments API, throwing ApiError on a non-OK status. */
async function send<T>(
  method: string,
  path: string,
  body?: unknown,
  signal?: AbortSignal,
): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method,
    credentials: 'same-origin',
    headers: body === undefined ? undefined : { 'Content-Type': 'application/json' },
    body: body === undefined ? undefined : JSON.stringify(body),
    signal,
  })
  if (!res.ok) {
    throw new ApiError(res.status, await readErrorMessage(res))
  }
  if (res.status === 204) {
    return undefined as T
  }
  const text = await res.text()
  return (text === '' ? undefined : JSON.parse(text)) as T
}

/**
 * Builds the thread path for a subject, escaping the uid. The two collections are
 * the only difference between a photo thread and a task thread.
 */
function threadPath(subject: CommentSubject): string {
  const collection = subject.kind === 'task' ? 'tasks' : 'photos'
  return `/${collection}/${encodeURIComponent(subject.uid)}/comments`
}

/**
 * Lists a subject's comments, **oldest first** — a conversation reads forwards.
 *
 * A subject with no comments (and, deliberately, one that does not exist) yields
 * an empty array rather than a 404, so an empty thread is a normal result.
 */
export async function fetchComments(
  subject: CommentSubject,
  signal?: AbortSignal,
): Promise<Comment[]> {
  const body = await send<CommentListResponse>('GET', threadPath(subject), undefined, signal)
  return body.comments
}

/**
 * Appends a comment to a subject's thread and returns the created record.
 *
 * @throws ApiError 400 (blank or over-long body), 404 (no such photo) or 429 (the
 *   per-user rate limit — the caller should say "slow down", not "it failed").
 */
export async function createComment(
  subject: CommentSubject,
  body: string,
  signal?: AbortSignal,
): Promise<Comment> {
  const created = await send<Comment>('POST', threadPath(subject), { body }, signal)
  announceTaskWrite(subject)
  return created
}

/**
 * Tells the task-change bus a task's thread was written to, so the counts
 * behind the navigation badge ("whose move is it") refresh at once. A photo's
 * thread is not the queue's business and stays silent.
 */
function announceTaskWrite(subject: CommentSubject): void {
  if (subject.kind === 'task') {
    notifyTaskChanged()
  }
}

/**
 * Rewrites the body of the caller's own comment and returns the edited record,
 * now carrying `edited_at`.
 *
 * @throws ApiError 403 (someone else's comment — admins included: an admin may
 *   remove a comment but never rewrite what someone is recorded as having said) or
 *   404 (already deleted, or addressed through the wrong subject).
 */
export async function updateComment(
  subject: CommentSubject,
  commentUid: string,
  body: string,
  signal?: AbortSignal,
): Promise<Comment> {
  const edited = await send<Comment>(
    'PATCH',
    `${threadPath(subject)}/${encodeURIComponent(commentUid)}`,
    { body },
    signal,
  )
  announceTaskWrite(subject)
  return edited
}

/**
 * Deletes a comment — the author's own, or (for an admin) anyone's. The delete is
 * soft server-side, so the comment simply drops out of every read.
 *
 * @throws ApiError 403 (not the author and not an admin) or 404 (already deleted).
 */
export async function deleteComment(
  subject: CommentSubject,
  commentUid: string,
  signal?: AbortSignal,
): Promise<void> {
  await send<undefined>(
    'DELETE',
    `${threadPath(subject)}/${encodeURIComponent(commentUid)}`,
    undefined,
    signal,
  )
  announceTaskWrite(subject)
}

File: internal/phototaskapi/comments.go
package phototaskapi

import (
	"errors"
	"fmt"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/panbotka/kukatko/internal/audit"
	"github.com/panbotka/kukatko/internal/auth"
	"github.com/panbotka/kukatko/internal/comments"
)

// commentRequest is the JSON body of the answer and edit endpoints: the
// plain-text body, trimmed and length-checked by the store.
type commentRequest struct {
	Body string `json:"body"`
}

// commentListResponse is the JSON body of the thread endpoint. Comments is
// always an array (never null) so a client can render an empty thread without a
// guard.
type commentListResponse struct {
	Comments []comments.Comment `json:"comments"`
}

// handleListComments returns a task's thread, oldest first, each comment with
// its author's resolved name. Every authenticated role may read it.
func (a *API) handleListComments(w http.ResponseWriter, r *http.Request) {
	if !a.commentsAvailable(w) {
		return
	}
	list, err := a.comments.List(r.Context(), comments.TaskSubject(taskUID(r)))
	if err != nil {
		writeCommentError(w, err, "listing comments failed")
		return
	}
	writeJSON(w, http.StatusOK, commentListResponse{Comments: list})
}

// handleCreateComment appends an answer to a task's thread and writes 201.
//
// It is guarded by RequireAuth rather than RequireWrite on purpose: answering is
// exactly what a viewer account is for here. The person who remembers the year a
// house was rebuilt is not the person who edits the library.
func (a *API) handleCreateComment(w http.ResponseWriter, r *http.Request) {
	if !a.commentsAvailable(w) {
		return
	}
	user, ok := currentUser(w, r)
	if !ok {
		return
	}
	var body commentRequest
	if err := decodeJSON(r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	uid := taskUID(r)
	created, err := a.comments.Create(r.Context(), comments.TaskSubject(uid), user.UID, body.Body,
		commentEntry(r, user.UID, audit.ActionCommentCreate, uid))
	if err != nil {
		writeCommentError(w, err, "creating comment failed")
		return
	}
	writeJSON(w, http.StatusCreated, created)
}

// handleUpdateComment rewrites the caller's own comment.
func (a *API) handleUpdateComment(w http.ResponseWriter, r *http.Request) {
	existing, user, ok := a.resolveComment(w, r)
	if !ok {
		return
	}
	if !canEditComment(existing, user) {
		writeError(w, http.StatusForbidden, "only the author may edit a comment")
		return
	}
	var body commentRequest
	if err := decodeJSON(r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	updated, err := a.comments.Update(r.Context(), existing.UID, body.Body,
		commentEntry(r, user.UID, audit.ActionCommentUpdate, existing.TaskUID))
	if err != nil {
		writeCommentError(w, err, "updating comment failed")
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

// handleDeleteComment soft-deletes the caller's own comment, or any comment when
// the caller is an admin, and writes 204.
func (a *API) handleDeleteComment(w http.ResponseWriter, r *http.Request) {
	existing, user, ok := a.resolveComment(w, r)
	if !ok {
		return
	}
	if !canDeleteComment(existing, user) {
		writeError(w, http.StatusForbidden, "only the author or an admin may delete a comment")
		return
	}
	if err := a.comments.Delete(r.Context(), existing.UID,
		commentEntry(r, user.UID, audit.ActionCommentDelete, existing.TaskUID)); err != nil {
		writeCommentError(w, err, "deleting comment failed")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// resolveComment reads the acting user and the {commentUID} comment, and refuses
// one that belongs to a different task with a 404 — so a comment UID cannot be
// probed through an unrelated task's thread.
func (a *API) resolveComment(w http.ResponseWriter, r *http.Request) (comments.Comment, auth.User, bool) {
	if !a.commentsAvailable(w) {
		return comments.Comment{}, auth.User{}, false
	}
	user, ok := currentUser(w, r)
	if !ok {
		return comments.Comment{}, auth.User{}, false
	}
	existing, err := a.comments.Get(r.Context(), chi.URLParam(r, "commentUID"))
	if err != nil {
		writeCommentError(w, err, "reading comment failed")
		return comments.Comment{}, auth.User{}, false
	}
	if existing.TaskUID != taskUID(r) {
		writeError(w, http.StatusNotFound, "comment not found")
		return comments.Comment{}, auth.User{}, false
	}
	return existing, user, true
}

// canEditComment reports whether user may rewrite the comment: only its author,
// and only when the comment still has one (an authorless comment, left behind by
// a deleted account, is editable by nobody).
func canEditComment(c comments.Comment, user auth.User) bool {
	return c.AuthorUID != "" && c.AuthorUID == user.UID
}

// canDeleteComment reports whether user may remove the comment: its author, or
// an admin removing anyone's.
func canDeleteComment(c comments.Comment, user auth.User) bool {
	return (c.AuthorUID != "" && c.AuthorUID == user.UID) || user.Role.IsAdmin()
}

// commentsAvailable reports whether a comment store is configured, writing 503
// when it is not. A task itself still reads and edits without one.
func (a *API) commentsAvailable(w http.ResponseWriter) bool {
	if a.comments == nil {
		writeError(w, http.StatusServiceUnavailable, "comments are unavailable")
		return false
	}
	return true
}

// commentEntry builds the audit entry for a thread mutation. The target is the
// task the comment hangs off; the store adds the comment's own UID to the
// details.
func commentEntry(r *http.Request, actorUID, action, taskUID string) audit.Entry {
	return audit.FromRequest(r, actorUID).Entry(action, "photo_tasks", taskUID, nil)
}

// writeCommentError maps a comments store error to an HTTP response: 404 for a
// missing task or comment, 400 for an invalid body, otherwise 500 with failMsg.
func writeCommentError(w http.ResponseWriter, err error, failMsg string) {
	switch {
	case errors.Is(err, comments.ErrSubjectNotFound):
		writeError(w, http.StatusNotFound, "task not found")
	case errors.Is(err, comments.ErrNotFound):
		writeError(w, http.StatusNotFound, "comment not found")
	case errors.Is(err, comments.ErrEmptyBody):
		writeError(w, http.StatusBadRequest, "comment body is empty")
	case errors.Is(err, comments.ErrBodyTooLong):
		writeError(w, http.StatusBadRequest,
			fmt.Sprintf("comment body is longer than %d characters", comments.MaxBodyLen))
	default:
		writeError(w, http.StatusInternalServerError, failMsg)
	}
}
